src/aiida/storage/sqlite_zip/migrator.py

```diff
@@ -68,6 +68,52 @@ def validate_storage(inpath: Path) -> None:
         )
 
 
+def check_migration_needed(inpath: Union[str, Path], target_version: str) -> bool:
+    """Check if migration is needed and validate the input file.
+
+    :param inpath: Path to the input file
+    :param target_version: Target version to migrate to
+    :return: True if migration is needed, False otherwise
+    :raises: StorageMigrationError if versions are invalid or file is corrupt
+    :raises: CorruptStorage if the file cannot be read properly
+    """
+    inpath = Path(inpath)
+
+    # the file should be either a tar (legacy only) or zip file
+    if not (tarfile.is_tarfile(str(inpath)) or zipfile.is_zipfile(str(inpath))):
+        msg = f'The input file is neither a tar nor a zip file: {inpath}'
+        raise CorruptStorage(msg)
+
+    # read the metadata.json which should always be present
+    metadata = extract_metadata(inpath, search_limit=None)
+
+    # obtain the current version from the metadata
+    if 'export_version' not in metadata:
+        msg = 'No export_version found in metadata'
+        raise CorruptStorage(msg)
+    current_version = metadata['export_version']
+
+    # check versions are valid
+    # versions 0.1, 0.2, 0.3 are no longer supported,
+    # since 0.3 -> 0.4 requires costly migrations of repo files (you would need to unpack all of them)
+    if current_version in ('0.1', '0.2', '0.3') or target_version in ('0.1', '0.2', '0.3'):
+        msg = (
+            f"Legacy migration from '{current_version}' -> '{target_version}' is not supported in aiida-core v2. "
+            'First migrate them to the latest version in aiida-core v1.'
+        )
+        raise StorageMigrationError(msg)
+    all_versions = list_versions()
+    if current_version not in all_versions:
+        msg = f"Unknown current version '{current_version}'"
+        raise StorageMigrationError(msg)
+    if target_version not in all_versions:
+        msg = f"Unknown target version '{target_version}'"
+        raise StorageMigrationError(msg)
+
+    # check if migration is needed
+    return current_version != target_version
```

Comment:
You can return the `current_version` and use it in the log.
```suggestion
    return current_version != target_version, current_version
```

Reply:
I tried this, but didn't like this approach because it muddied the return of the function, and in most cases, I ended up just throwing the `current_version` return away (apart from logging). This led to a general refactor now, where I split the `check_migration_needed` code into two functions, `get_current_archive_version` and `validate_archive_versions`, both as `staticmethods` of the `SqliteZipBackend`. I feel like this is more readable. Happy to read your thoughts :)